File: LeetCodeCPP/443.string-compression.cpp
/*
 * @lc app=leetcode id=443 lang=cpp
 *
 * [443] String Compression
 */

#include <vector>
#include <string>

using namespace std;

// @lc code=start
class Solution
{
public:
    int compress(vector<char> &chars)
    {
        int groupIdx = 0;
        char currentChar = chars[0];
        int count = 1;

        for (size_t i = 1; i < chars.size(); i++)
        {
            if (chars[i] == currentChar)
            {
                count++;
            }
            else
            {
                if (count > 1)
                {
                    string count_str = to_string(count);
                    for (size_t j = 0; j < count_str.length(); j++)
                    {
                        chars[++groupIdx] = count_str[j];
                    }
                    chars[++groupIdx] = chars[i];
                    currentChar = chars[i];
                    count = 1;
                }
                else
                {
                    chars[++groupIdx] = chars[i];
                    currentChar = chars[i];
                    // groupIdx++;
                    count = 1;
                }
            }
        }

        if (count > 1)
        {
            string count_str = to_string(count);
            for (size_t j = 0; j < count_str.length(); j++)
            {
                chars[++groupIdx] = count_str[j];
            }
        }

        return groupIdx + 1;
    }
};
// @lc code=end

File: LeetCodePython/443.string-compression.py
#
# @lc app=leetcode id=443 lang=python3
#
# [443] String Compression
#

from typing import List


# @lc code=start
class Solution:
    def compress(self, chars: List[str]) -> int:
        group_idx = 0
        currentChar = chars[0]
        count = 1

        for i in range(1, len(chars)):
            if chars[i] == currentChar:
                count += 1
            else:
                if count > 1:
                    count_str = str(count)
                    for j in range(len(count_str)):
                        group_idx += 1
                        chars[group_idx] = count_str[j]

                group_idx += 1
                chars[group_idx] = chars[i]
                currentChar = chars[i]
                count = 1

        if count > 1:
            count_str = str(count)
            for j in range(len(count_str)):
                group_idx += 1
                chars[group_idx] = count_str[j]

        return group_idx + 1


# @lc code=end
